Add wavelet loss

rockerBOO opened the pull request:
Some papers about wavelets: 
https://arxiv.org/abs/2306.00306
https://arxiv.org/abs/2402.19215
https://arxiv.org/abs/2211.16152
https://arxiv.org/abs/2407.12538

Video about Wavelets
https://www.youtube.com/watch?v=jnxqHcObNK4


Activate Wavelet Loss: 
```toml
wavelet_loss = true
```

Configure Wavelet loss: 

```toml
wavelet_loss_type = "huber" # l2, l1, smooth_l2, huber
wavelet_loss_transform = "swt" # dwt, swt
wavelet_loss_wavelet = "haar" # over 100 wavelets, but try db4, sym7. AI toolkit uses haar. 
wavelet_loss_level = 2 # Loss level is how many levels we process. DWT maybe 1-3, SWT 1-2. SWT level 2 focuses on details more. 
wavelet_loss_alpha = 0.01 # How much impact the loss has on the latent loss value
```

We use a custom loss implementation and I just learned about Wavelets like yesterday so we may consider how the approach works and maybe utilize other libraries as well.

recris commented on train_network.py:
```diff
@@ -462,6 +463,18 @@ def process_batch(
+
+            wav_loss, pred_combined_hf, target_combined_hf = self.wavelet_loss(model_denoised.float(), flow_based_clean.float())
+            # Weight the losses as needed
+            loss = loss + args.wavelet_loss_alpha * wav_loss
```
Maybe do `loss = (1.0 - args.wavelet_loss_alpha) * loss + args.wavelet_loss_alpha * wav_loss` so its a proper interpolation?

recris commented:
You should consider including a weighting/masking scheme for the different levels, I am getting **amazing** results from it.

I've been playing with a prototype of this myself, see https://github.com/kohya-ss/sd-scripts/discussions/294#discussioncomment-12710624

For example, masking the lowpass elements makes it easier to learn subjects and objects without transferring the overall image aesthetic bias.

Here is my hacky training code as an example:
```
use_wavelet_loss = True
        wavelet_loss_ratio = 0.98

        huber_c = train_util.get_huber_threshold_if_needed(args, timesteps, noise_scheduler)
        loss = train_util.conditional_loss(noise_pred.float(), target.float(), args.loss_type, "none", huber_c)

        if args.masked_loss or ("alpha_masks" in batch and batch["alpha_masks"] is not None):
            # not applying the mask by multiplying the loss, instead this scales the gradients directly
            # it enables using other loss functions below without caring about having to worry
            # about mask being incompatible
             loss, mask = apply_masked_loss(loss, batch)
            mask = torch.nn.functional.interpolate(mask, size=noise_pred.shape[2:], mode="area")
            noise_pred.register_hook(lambda grad: grad * mask.to(grad.dtype))

        loss = loss.mean([1, 2, 3])

        if use_wavelet_loss:
            # custom weighting scheme, this should be parameter
            num_levels = 3
            level_weights = [1.0] * num_levels
            lowpass_weight = 0.0
            level_weights[1] = 0.0
            level_weights[2] = 0.0

            # this requires pytorch-wavelets to be installed
            dwt = DWTForward(J=num_levels, mode='zero', wave='haar').to(device=accelerator.device, dtype=vae_dtype)

            model_pred_xl, model_pred_xh = dwt(noise_pred)
            target_xl, target_xh = dwt(target)

            # compute lowpass loss
            wt_loss = train_util.conditional_loss(model_pred_xl.float(), target_xl.float(), args.loss_type, "none", huber_c)
            wt_loss = wt_loss * lowpass_weight

            # compute loss for each band
            for lvl, (p, t) in enumerate(zip(reversed(model_pred_xh), reversed(target_xh))):
                l = train_util.conditional_loss(p.float(), t.float(), args.loss_type, "none", huber_c)
                l = l * level_weights[lvl]

                l_xlh, l_xhl, l_xhh = torch.unbind(l, dim=2)

                wt_loss = torch.cat((
                    torch.cat((wt_loss, l_xlh), dim=3),
                    torch.cat((l_xhl, l_xhh), dim=3)),
                    dim=2)

            wt_loss = wt_loss.mean([1, 2, 3])

            loss = wavelet_loss_ratio * wt_loss + (1 - wavelet_loss_ratio) * loss
```

rockerBOO commented:
@recris I went through and added inputs for band weighting, allowing a good amount of customization. It was in there previously for SWT but now it is applied to both. The default weights are low so as to allow one to customize how much impact each band has. The alpha is now more like a multiplier instead but should control the weighting via the bands ideally. 

I added wavelet loss to the logging, which can help when adjusting the band weights to see the wavelet impacts. 
![Screenshot 2025-04-12 at 02-18-58 women-flux-kohya-lora Workspace – Weights   Biases](https://github.com/user-attachments/assets/09aed361-e4c0-40c5-9fa5-5df9125db872)

| Example | Example |
|--------|--------|
| ![c-f1-2025-04-12_00059_](https://github.com/user-attachments/assets/6dcee734-d9ea-4915-b68c-fe5dbdbd37ac) | ![c-f1-2025-04-12_00058_](https://github.com/user-attachments/assets/b51df9a4-c664-43d6-9afb-8bae3bb228a3) | 
| ![c-f1-2025-04-12_00028_](https://github.com/user-attachments/assets/bde0315a-2479-4662-8122-7a37f4908faf) | ![c-f1-2025-04-12_00027_](https://github.com/user-attachments/assets/419c071b-f7aa-4ab4-b41f-4a89995c4d2b) |
| ![c-f1-2025-04-12_00008_](https://github.com/user-attachments/assets/49f3f135-d9aa-4e61-b614-fd150923ea06) | ![c-f1-2025-04-12_00024_](https://github.com/user-attachments/assets/32a739d8-49f1-4b5a-bede-0696eda8cb9c) |


I am still working through some trainings of this to get some more examples and comparisons.

EClipXAi commented:
Traceback (most recent call last):
  File "/workspace/kohya_ss/sd-scripts/flux_train_network.py", line 559, in <module>
    trainer.train(args)
  File "/workspace/kohya_ss/sd-scripts/train_network.py", line 1473, in train
    loss, wav_loss = self.process_batch(
  File "/workspace/kohya_ss/sd-scripts/train_network.py", line 496, in process_batch
    wav_loss, pred_combined_hf, target_combined_hf = self.wavelet_loss(model_denoised.float(), flow_based_clean.float())
  File "/workspace/kohya_ss/venv/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1736, in _wrapped_call_impl
    return self._call_impl(*args, **kwargs)
  File "/workspace/kohya_ss/venv/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1747, in _call_impl
    return forward_call(*args, **kwargs)
  File "/workspace/kohya_ss/sd-scripts/library/custom_train_functions.py", line 804, in forward
    band_loss = self.band_level_weights.get(weight_key, self.band_weights[band]) * self.loss_fn(pred_stack, target_stack)
  File "/workspace/kohya_ss/sd-scripts/train_network.py", line 488, in loss_fn
    huber_c = train_util.get_huber_threshold_if_needed(args, timesteps, latents, noise_scheduler)
TypeError: get_huber_threshold_if_needed() takes 3 positional arguments but 4 were given
Traceback (most recent call last):
  File "/workspace/kohya_ss/sd-scripts/flux_train_network.py", line 559, in <module>
    trainer.train(args)
  File "/workspace/kohya_ss/sd-scripts/train_network.py", line 1473, in train
    loss, wav_loss = self.process_batch(
  File "/workspace/kohya_ss/sd-scripts/train_network.py", line 496, in process_batch
    wav_loss, pred_combined_hf, target_combined_hf = self.wavelet_loss(model_denoised.float(), flow_based_clean.float())
  File "/workspace/kohya_ss/venv/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1736, in _wrapped_call_impl
    return self._call_impl(*args, **kwargs)
  File "/workspace/kohya_ss/venv/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1747, in _call_impl
    return forward_call(*args, **kwargs)
  File "/workspace/kohya_ss/sd-scripts/library/custom_train_functions.py", line 804, in forward
    band_loss = self.band_level_weights.get(weight_key, self.band_weights[band]) * self.loss_fn(pred_stack, target_stack)
  File "/workspace/kohya_ss/sd-scripts/train_network.py", line 488, in loss_fn
    huber_c = train_util.get_huber_threshold_if_needed(args, timesteps, latents, noise_scheduler)
TypeError: get_huber_threshold_if_needed() takes 3 positional arguments but 4 were given


getting this error when trying it

rockerBOO commented:
@EClipXAi Apologies, some newer functionality leaked in there. I have fixed this. Give it another shot and let me know if you have any issues.

rockerBOO commented:
Added QuaterionWaveletTransform derived from https://arxiv.org/abs/2505.00334. 

`transform_type = "qwt"`

```
wavelet_loss = true
wavelet_loss_type = "huber"
wavelet_loss_transform = "qwt"
wavelet_loss_wavelet = "sym7"
wavelet_loss_level = 3
wavelet_loss_alpha = 1
wavelet_loss_band_weights = { "ll" = 0.25, "lh" = 1.0, "hl" = 1.0, "hh" = 1.0 }
wavelet_loss_quaternion_component_weights = { "r" = 0.25, "i" = 0.5, "j" = 0.5, "k" = 0.5 }
wavelet_loss_ll_level_threshold = -1
```

It should work similarly to DWT but with more components (4 vs 1 with DWT) using hilbert filters. It's probably experimental so true recommended values I'm not quite sure yet. 

Additionally reworked SWT to work better, and SWT should be more performant now using 1d convolutions.

`transform_type` wasn't properly being used before so now swt and qwt should work where it was only using DWT before.

Added tests for the Wavelet tranforms and Wavelet loss to make sure it's working as expected.

rockerBOO commented:
@WinodePino During the startup process there will be a section that is like

```
Wavelet Loss:
	Level: 2
	Alpha: 1.0
	Transform: swt
	Wavelet: sym7
	LL level threshold: -1
```

rockerBOO commented:
> Thanks! I found the issue, but now I run into this:
> 
> `AttributeError: module 'torch' has no attribute 'float8_e4m3fnuz'`
> 
> If I start a training with the normal SD3 branch this error won't show up.

Do you have all of the error? Also which pytorch version are using?

rockerBOO commented:
it's saying `torch.float8_e4m3fnuz` isn't a attribute which is likely a pytorch version issue. With the venv enabled do 

```
python -c "import torch; print(torch.__version__)"
```

Otherwise it's a bigger issue.

WinodePino commented:
> it's saying `torch.float8_e4m3fnuz` isn't a attribute which is likely a pytorch version issue. With the venv enabled do
> 
> ```
> python -c "import torch; print(torch.__version__)"
> ```
> 
> Otherwise it's a bigger issue.

Didn't work. Because my normal SD Scripts did work I tried to manually replace all necessary changed files and now it starts the training and it shows the section you mentioned during start up.

Although I was expecting training to be a lot slower, but it seem to be similar with 4s/it on a 4090. So I am not sure if it is working as expected.

rockerBOO commented:
Added new loss calculation

WinodePino commented:
Should wavelet_loss_type replace loss_type and should I remove it or not? My results so far are letting me generate on higher resolution and it looks nice, but I expected it to be good in the details (creating automotive lora's), but somehow I got way better details with normal LoRa training on a dedistilled model. Am I missing something? I am training on adafactor, 128 rank/alpha. 

`loss_type = "l2"
wavelet_loss = true
wavelet_loss_metrics = true
wavelet_loss_type = "l1" # l2, l1, smooth_l2, huber. l1 and huber would be recommended
wavelet_loss_transform = "swt" # dwt, swt. swt keeps the spatial details but dwt can be a little more efficient.
wavelet_loss_wavelet = "sym7" # over 100 wavelets, but try db4, sym7. AI toolkit uses haar. 
wavelet_loss_level = 2 # Loss level is how many levels we process. DWT maybe 1-3, SWT 1-2. SWT level 2 focuses on details more. 
wavelet_loss_alpha = 1.0 # How much impact the loss has on the latent loss value
wavelet_loss_band_level_weights = { "ll1" = 0.1, "lh1" = 0.01, "hl1" = 0.01, "hh1" = 0.05, "ll2" = 0.1, "lh2" = 0.01, "hl2" = 0.01, "hh2" = 0.05 } # Set the individual levels band weights. Starts at level 1
wavelet_loss_band_weights = { "ll" = 0.1, "lh" = 0.01, "hl" = 0.01, "hh" = 0.05 } # Sets the defaults for the bands. Currently need to set all the values or bad things might happen.
wavelet_loss_ll_level_threshold = -1 # level to process the ll at. Low frequency will be similar to the original latent so only need the last levels for that detail.
wavelet_loss_rectified_flow = true # Experimental. Not recommended to change, but toggles rectified flow to get clean latents.`

deGENERATIVE-SQUAD commented:
Found this PR finally. I hope you don’t mind if I share something relevant to the topic.

A guy from an one forum made this wavelet implementation based on many tests (according to his speech) of different configurations. 

`@torch.compile(dynamic=True)
def compute_weights(timestep, width, shift):
    timestep = min(1000, 1000 - timestep + shift)
    x = torch.linspace(0, 1000, 1000).cuda()
    func = lambda x, mu, sigma, amplitude: amplitude * torch.exp(-(x - mu)**2 / (2 * sigma**2))
    y = func(x, mu=timestep, sigma=width, amplitude=1.0)
    weights = [y[i].item() for i in [0, 200, 400, 600, 800, 999]]
    return weights


@torch.compile(dynamic=True)
def compute_loss(model_pred, target, timesteps):
    loss_levels = []
    for b in range(model_pred.shape[0]):
        pred = model_pred[b, ...].unsqueeze(0)
        tgt = target[b, ...].unsqueeze(0)
        num_levels = math.ceil(math.log2(max(pred.shape[2], pred.shape[3])))
        dwt = DWTForward(J=num_levels, mode='zero', wave="haar").to(device=pred.device, dtype=torch.float)

        model_pred_xl, model_pred_xh = dwt(pred.float())
        target_xl, target_xh = dwt(tgt.float())
        model_pred_l0 = model_pred_xl.unsqueeze(2)
        target_l0 = target_xl.unsqueeze(2)

        level_weights = compute_weights(timesteps[b], 350, 0)
        for p, t, w in zip(model_pred_xh + [model_pred_l0], target_xh + [target_l0], range(1, num_levels + 1)):
            l = F.mse_loss(p.float(), t.float(), reduction="none")
            loss_levels.append(level_weights[min(num_levels + 1 - w, len(level_weights) - 1)] * l.mean([2,3,4]) / float(2 ** w))

    loss = torch.stack(loss_levels, dim=2).mean([1, 2]).mean()
    return loss
`

For those who want to manually integrate this into sd-scripts and aren’t quite sure how, the monkey patch is:

In train_network:

`huber_c = train_util.get_huber_threshold_if_needed(args, timesteps, noise_scheduler)
        loss = train_util.conditional_loss(noise_pred.float(), target.float(), args.loss_type, "none", huber_c, timesteps)
        if weighting is not None:
            loss = loss * weighting
        if args.masked_loss or ("alpha_masks" in batch and batch["alpha_masks"] is not None):
            loss = apply_masked_loss(loss, batch)
        loss = loss`

In train_util:

a) Define a new loss type:

`def conditional_loss(
    model_pred: torch.Tensor, target: torch.Tensor, loss_type: str, reduction: str, huber_c: Optional[torch.Tensor] = None, timesteps: Optional[torch.Tensor] = None
):
    """
    NOTE: if you're using the scheduled version, huber_c has to depend on the timesteps already
    """
    if loss_type == "l2":
        loss = torch.nn.functional.mse_loss(model_pred, target, reduction=reduction)
    elif loss_type == "dwt":
        if timesteps is None:
            raise ValueError("Wavelet loss requires `timesteps`")
        loss_levels = []
        for b in range(model_pred.shape[0]):
            pred = model_pred[b, ...].unsqueeze(0)
            tgt = target[b, ...].unsqueeze(0)
            num_levels = math.ceil(math.log2(max(pred.shape[2], pred.shape[3])))
            dwt = DWTForward(J=num_levels, mode='zero', wave="haar").to(device=pred.device, dtype=torch.float)

            model_pred_xl, model_pred_xh = dwt(pred.float())
            target_xl, target_xh = dwt(tgt.float())
            model_pred_l0 = model_pred_xl.unsqueeze(2)
            target_l0 = target_xl.unsqueeze(2)

            level_weights = compute_weights(timesteps[b], 350, 0)
            for p, t, w in zip(model_pred_xh + [model_pred_l0], target_xh + [target_l0], range(1, num_levels + 1)):
                l = F.mse_loss(p.float(), t.float(), reduction="none")
                weight_idx = min(num_levels + 1 - w, len(level_weights) - 1)
                loss_levels.append(level_weights[weight_idx] * l.mean([2, 3, 4]))

        loss = torch.stack(loss_levels, dim=0).mean()
    elif loss_type == "l1":
        loss = torch.nn.functional.l1_loss(model_pred, target, reduction=reduction)
    elif loss_type == "huber":
        if huber_c is None:
            raise NotImplementedError("huber_c not implemented correctly")
        huber_c = huber_c.view(-1, 1, 1, 1)
        loss = 2 * huber_c * (torch.sqrt((model_pred - target) ** 2 + huber_c**2) - huber_c)
        if reduction == "mean":
            loss = torch.mean(loss)
        elif reduction == "sum":
            loss = torch.sum(loss)
    elif loss_type == "smooth_l1":
        if huber_c is None:
            raise NotImplementedError("huber_c not implemented correctly")
        huber_c = huber_c.view(-1, 1, 1, 1)
        loss = 2 * (torch.sqrt((model_pred - target) ** 2 + huber_c**2) - huber_c)
        if reduction == "mean":
            loss = torch.mean(loss)
        elif reduction == "sum":
            loss = torch.sum(loss)
    else:
        raise NotImplementedError(f"Unsupported Loss Type: {loss_type}")
    return loss`
    
Define arg:

`parser.add_argument(
        "--loss_type",
        type=str,
        default="l2",
        choices=["l1", "l2", "huber", "smooth_l1", "dwt"],
        help="The type of loss function to use (L1, L2, Huber, or smooth L1), default is L2 / 使用する損失関数の種類（L1、L2、Huber、またはsmooth L1）、デフォルトはL2",
    )`

b) After the imports, insert:

`def compute_weights(timestep, width, shift):
    timestep = min(1000, 1000 - timestep + shift)
    x = torch.linspace(0, 1000, 1000).cuda()
    func = lambda x, mu, sigma, amplitude: amplitude * torch.exp(-(x - mu)**2 / (2 * sigma**2))
    y = func(x, mu=timestep, sigma=width, amplitude=1.0)
    weights = [y[i].item() for i in [0, 200, 400, 600, 800, 999]]
    return weights`

c) Add to the imports:
import torch.nn.functional as F  
from pytorch_wavelets import DWTForward

I’ve been experimenting with different optimizers for a week now, and the quality achieved with this implementation is simply excellent. Far better than l2/l2/huber. It works fine with --debiased_estimation_loss, by the way.

WinodePino commented:
> Found this PR finally. I hope you don’t mind if I share something relevant to the topic.
> 
> A guy from an one forum made this wavelet implementation based on many tests (according to his speech) of different configurations.
> 
> ```
> @torch.compile(dynamic=True)
> def compute_weights(timestep, width, shift):
>     timestep = min(1000, 1000 - timestep + shift)
>     x = torch.linspace(0, 1000, 1000).cuda()
>     func = lambda x, mu, sigma, amplitude: amplitude * torch.exp(-(x - mu)**2 / (2 * sigma**2))
>     y = func(x, mu=timestep, sigma=width, amplitude=1.0)
>     weights = [y[i].item() for i in [0, 200, 400, 600, 800, 999]]
>     return weights`
> @torch.compile(dynamic=True)`
> def compute_loss(model_pred, target, timesteps):
>     loss_levels = []
>     for b in range(model_pred.shape[0]):
>         pred = model_pred[b, ...].unsqueeze(0)
>         tgt = target[b, ...].unsqueeze(0)
>         num_levels = math.ceil(math.log2(max(pred.shape[2], pred.shape[3])))
>         dwt = DWTForward(J=num_levels, mode='zero', wave="haar").to(device=pred.device, dtype=torch.float)`
> 
>         model_pred_xl, model_pred_xh = dwt(pred.float())
>         target_xl, target_xh = dwt(tgt.float())
>         model_pred_l0 = model_pred_xl.unsqueeze(2)
>         target_l0 = target_xl.unsqueeze(2)
> 
>         level_weights = compute_weights(timesteps[b], 350, 0)
>         for p, t, w in zip(model_pred_xh + [model_pred_l0], target_xh + [target_l0], range(1, num_levels + 1)):
>             l = F.mse_loss(p.float(), t.float(), reduction="none")
>             loss_levels.append(level_weights[min(num_levels + 1 - w, len(level_weights) - 1)] * l.mean([2,3,4]) / float(2 ** w))
> 
>     loss = torch.stack(loss_levels, dim=2).mean([1, 2]).mean()
>     return loss
> ```
> 
> For those who want to manually integrate this into sd-scripts and aren’t quite sure how, the monkey patch is:
> 
> In train_network:
> 
> ```
> huber_c = train_util.get_huber_threshold_if_needed(args, timesteps, noise_scheduler)
>         loss = train_util.conditional_loss(noise_pred.float(), target.float(), args.loss_type, "none", huber_c, timesteps)
>         if weighting is not None:
>             loss = loss * weighting
>         if args.masked_loss or ("alpha_masks" in batch and batch["alpha_masks"] is not None):
>             loss = apply_masked_loss(loss, batch)
>         loss = loss
> ```
> 
> In train_util:
> 
> a) Define a new loss type:
> 
> ```
> def conditional_loss(
>     model_pred: torch.Tensor, target: torch.Tensor, loss_type: str, reduction: str, huber_c: Optional[torch.Tensor] = None, timesteps: Optional[torch.Tensor] = None
> ):
>     """
>     NOTE: if you're using the scheduled version, huber_c has to depend on the timesteps already
>     """
>     if loss_type == "l2":
>         loss = torch.nn.functional.mse_loss(model_pred, target, reduction=reduction)
>     elif loss_type == "dwt":
>         if timesteps is None:
>             raise ValueError("Wavelet loss requires `timesteps`")
>         loss_levels = []
>         for b in range(model_pred.shape[0]):
>             pred = model_pred[b, ...].unsqueeze(0)
>             tgt = target[b, ...].unsqueeze(0)
>             num_levels = math.ceil(math.log2(max(pred.shape[2], pred.shape[3])))
>             dwt = DWTForward(J=num_levels, mode='zero', wave="haar").to(device=pred.device, dtype=torch.float)
> 
>             model_pred_xl, model_pred_xh = dwt(pred.float())
>             target_xl, target_xh = dwt(tgt.float())
>             model_pred_l0 = model_pred_xl.unsqueeze(2)
>             target_l0 = target_xl.unsqueeze(2)
> 
>             level_weights = compute_weights(timesteps[b], 350, 0)
>             for p, t, w in zip(model_pred_xh + [model_pred_l0], target_xh + [target_l0], range(1, num_levels + 1)):
>                 l = F.mse_loss(p.float(), t.float(), reduction="none")
>                 weight_idx = min(num_levels + 1 - w, len(level_weights) - 1)
>                 loss_levels.append(level_weights[weight_idx] * l.mean([2, 3, 4]))
> 
>         loss = torch.stack(loss_levels, dim=0).mean()
>     elif loss_type == "l1":
>         loss = torch.nn.functional.l1_loss(model_pred, target, reduction=reduction)
>     elif loss_type == "huber":
>         if huber_c is None:
>             raise NotImplementedError("huber_c not implemented correctly")
>         huber_c = huber_c.view(-1, 1, 1, 1)
>         loss = 2 * huber_c * (torch.sqrt((model_pred - target) ** 2 + huber_c**2) - huber_c)
>         if reduction == "mean":
>             loss = torch.mean(loss)
>         elif reduction == "sum":
>             loss = torch.sum(loss)
>     elif loss_type == "smooth_l1":
>         if huber_c is None:
>             raise NotImplementedError("huber_c not implemented correctly")
>         huber_c = huber_c.view(-1, 1, 1, 1)
>         loss = 2 * (torch.sqrt((model_pred - target) ** 2 + huber_c**2) - huber_c)
>         if reduction == "mean":
>             loss = torch.mean(loss)
>         elif reduction == "sum":
>             loss = torch.sum(loss)
>     else:
>         raise NotImplementedError(f"Unsupported Loss Type: {loss_type}")
>     return loss
> ```
> 
> Define arg:
> 
> ```
> parser.add_argument(
>         "--loss_type",
>         type=str,
>         default="l2",
>         choices=["l1", "l2", "huber", "smooth_l1", "dwt"],
>         help="The type of loss function to use (L1, L2, Huber, or smooth L1), default is L2 / 使用する損失関数の種類（L1、L2、Huber、またはsmooth L1）、デフォルトはL2",
>     )
> ```
> 
> b) After the imports, insert:
> 
> ```
> def compute_weights(timestep, width, shift):
>     timestep = min(1000, 1000 - timestep + shift)
>     x = torch.linspace(0, 1000, 1000).cuda()
>     func = lambda x, mu, sigma, amplitude: amplitude * torch.exp(-(x - mu)**2 / (2 * sigma**2))
>     y = func(x, mu=timestep, sigma=width, amplitude=1.0)
>     weights = [y[i].item() for i in [0, 200, 400, 600, 800, 999]]
>     return weights
> ```
> 
> c) Add to the imports:
> 
> ```
> import torch.nn.functional as F  
> from pytorch_wavelets import DWTForward
> ```
> 
> I’ve been experimenting with different optimizers for a week now, and the quality achieved with this implementation is simply excellent. Far better than l1/l2/huber. It works fine with --debiased_estimation_loss, by the way.

I am trying to test this, but somehow my caching latents gets stuck at 0%, any idea why this could be?

And should it work with --loss_type dwt or should I also add something to the config.toml like wavelet_loss = true?

Thanks!

deGENERATIVE-SQUAD commented:
@WinodePino 

>Thanks, it is working now, but it doesnt seem to learn anything.

If the iteration is proceeding at a normal speed, then obviously the problem lies in the too low feature update speed.
Show your full configuration, please.

iqddd commented:
@rockerBOO Does it make sense to use it with Flux?

rockerBOO commented:
> > > @rockerBOO Does it make sense to use it with Flux?
> > 
> > 
> > yeah I started with Flux for this. I did get it "working" for diffusion UNet models (SD/SDXL) but it requires estimating the resulting end step latents so doesn't work as well as rectified flow models like Flux.
> 
> Hmm... tried it but complains about missing Huber loss implementation. FLUX isn't really a noise model tho, so kinda unclear.

```
wavelet_loss_type = "huber" # l2, l1, smooth_l2, huber. l1 and huber would be recommended
```

Use l1 instead here. I have another PR that adds huber support for flux.
